File: src/algorithms/binary_search.rs



pub fn execute(vector: &Vec<u32>, find: u32) -> (usize, u32, u32) {

  let mut min: u32 = 0;
  let mut max: u32 = (vector.len() - 1) as u32;
  let mut rounds: u32 = 0;

  while min <= max {
    let mid = f64::from((min + max) / 2).floor() as usize;

    if vector[mid] == find {
      return (mid, rounds, vector[mid]);
    }

    if vector[mid] < find {
      min = (mid as u32) + 1;
    }
    else {
      max = (mid as u32) - 1;
    }

    rounds += 1;
  }

  (0, 0, 0)
}

File: src/algorithms/binary_search.cpp
#include <cmath>
#include <vector>


#include "binary_search.h"


namespace binary_search {


  Result execute(
    const std::vector<unsigned int> &unsorted,
    unsigned const int search
  ) {

    unsigned int min = 0;
    unsigned int max = int(unsorted.size() - 1);
    unsigned int mid = 0;
    unsigned int rounds = 0;

    while (min <= max) {
      mid = int(std::floor((min + max) / 2));

      if (unsorted[mid] == search) {
        return { true, mid, rounds };
      }

      if (unsorted[mid] < search) {
        min = mid + 1;
      }
      else {
        max = mid - 1;
      }

      rounds++;
    }

    return { false, 0, rounds };
  }
}
